### PR TITLE
Allow underscores as separators in integers and floating point numbers

### PR DESCRIPTION
**Pull request by daviditen**
* Update chapel.lex to allow binary, octal, decimal, and hexadecimal
  integers to have underscores in them.
* Update chapel.lex to allow floating point numbers to have
  underscores.
* Update `buildIntLiteral` to remove the underscores.
* Update `new_FloatSymbol` to remove the underscores.
* Add tests for numbers with underscores.
* Update the "Lexical Structure" spec section.

TODO:
- [x] Add separators for binary, octal, hexadecimal literals
- [x] Add separators for floating point literals
- [x] Update the spec
- [ ] Update syntax highlighters
  - [x] Opened issue #11518 requesting this in a separate PR.

### CODE REVIEW

**Comment by bradcray** [comment]
This PR is going to need spec and (shudder) syntax highlighting updates as well.

File: compiler/AST/build.cpp [hunk 1]
```diff
@@ -264,8 +264,34 @@ Expr* buildIntLiteral(const char* pch, const char* file, int line) {
...
+      if (pch[i] == '_') {
+        if (j == 0 || j % 3 != 0) {
+          VarSymbol* temp = createASTforLineNumber(file, line);
+          USR_FATAL(temp, "underscores in integer literals are allowed "
+                          "only after groups of three digits");
+        }
+      } else {
+        j++;
+      }
+    }
```
Comment (mppf) [inline compiler/AST/build.cpp]: We decided to remove this check, right?

File: compiler/AST/build.cpp [hunk 2]
```diff
@@ -264,8 +264,34 @@ Expr* buildIntLiteral(const char* pch, const char* file, int line) {
     ull = octStr2uint64(pch, true, file, line);
   else if (!strncmp("0x", pch, 2) || !strncmp("0X", pch, 2))
     ull = hexStr2uint64(pch, true, file, line);
-  else
-    ull = str2uint64(pch, true, file, line);
+  else {
+    int j = 0;
+    int len = strlen(pch);
+    char noUnderscores[len+1];
```
Comment (mppf) [inline compiler/AST/build.cpp]: I can never remember if C run-time length arrays are legal in standard C or if they are an extension. In any case I'd probably just malloc the result here. If you keep the variable length array, you really really need to check that len isn't say 2GB.

File: compiler/parser/chapel.lex [hunk 3]
```diff
@@ -91,8 +91,9 @@ ident            {letter}({letter}|{digit}|"$")*
 
 binaryLiteral    0[bB]{bit}+
 octalLiteral     0[oO]{octDigit}+
+decimalLiteral   {digit}(_|{digit})*
 hexLiteral       0[xX]{hexDigit}+
-intLiteral       {digit}+|{binaryLiteral}|{octalLiteral}|{hexLiteral}
+intLiteral       {binaryLiteral}|{octalLiteral}|{decimalLiteral}|{hexLiteral}
```
Comment (mppf) [inline compiler/parser/chapel.lex]: I think this lexer change is missing the corresponding change for floating point literals.

File: compiler/AST/build.cpp (hunk 1 above)
Comment (daviditen) [reply compiler/AST/build.cpp]: Right.  I just pushed some new commits that include removing this check.

File: compiler/AST/build.cpp (hunk 2 above)
Comment (daviditen) [reply compiler/AST/build.cpp]: Good call. I just updated it to use malloc/free instead.

File: compiler/parser/chapel.lex (hunk 3 above)
Comment (daviditen) [reply compiler/parser/chapel.lex]: I've now added support for binary, octal, hexadecimal, and floating point literals.

**Comment by daviditen** [comment]
Passed Linux64 paratest.

File: compiler/parser/chapel.lex [hunk 4]
```diff
@@ -89,21 +89,24 @@ letter           [_a-zA-Z]
...
-floatLiteral3    {digit}+{exponent}
+binaryLiteral    0[bB]{bit}(_|{bit})*
+octalLiteral     0[oO]{octDigit}(_|{octDigit})*
+decimalLiteral   {digit}(_|{digit})*
+hexLiteral       0[xX]{hexDigit}(_|{hexDigit})*
+intLiteral       {binaryLiteral}|{octalLiteral}|{decimalLiteral}|{hexLiteral}
+
+digitsOrSeps     {digit}(_|{digit})*
+exponent         [Ee][\+\-]?{digitsOrSeps}
+floatLiteral1    ({digit}*|{digitsOrSeps})"."{digitsOrSeps}({exponent})?
```
Comment (mppf) [inline compiler/parser/chapel.lex]: I think this `({digit}*|{digitsOrSeps})` would be more reasonably expressed as `(|{digitsOrSeps})` but if that doesn't work for some reason what you have is OK.

Comment (daviditen) [reply compiler/parser/chapel.lex]: Now that you've suggested that this line could be improved, I think even better would be `{digitsOrSeps}?`.  I see that there are a couple unnecessary (but not outright harmful either) `+` operators in there too that I can clean up. I'll go ahead and do that and do another round of testing.

Comment (daviditen) [reply compiler/parser/chapel.lex]: Another round of testing has passed, so I think this is all set.  Thanks @mppf!

**Pull request merged**